builder/dockerfile/instructions/commands.go

```diff
@@ -0,0 +1,297 @@
+package instructions
+
+import (
+	"errors"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/strslice"
+)
+
+type KeyValuePair struct {
+	Key   string
+	Value string
+}
+
+func (kvp *KeyValuePair) String() string {
+	return kvp.Key + "=" + kvp.Value
+}
+
+type KeyValuePairs []KeyValuePair
+
+type BaseCommand struct {
+	code string
+	name string
+}
+
+func (c *BaseCommand) String() string {
+	return c.code
+}
+func (c *BaseCommand) Name() string {
+	return c.name
+}
+
+func newBaseCommand(req parseRequest) BaseCommand {
+	return BaseCommand{code: req.original, name: req.command}
+}
+
+// SingleWordExpander is a provider for variable expansion where 1 word => 1 output
+type SingleWordExpander func(word string) (string, error)
+
+type SupportsSingleWordExpansion interface {
+	Expand(expander SingleWordExpander) error
+}
+
+func expandKvp(kvp KeyValuePair, expander SingleWordExpander) (KeyValuePair, error) {
+	key, err := expander(kvp.Key)
+	if err != nil {
+		return KeyValuePair{}, err
+	}
+	value, err := expander(kvp.Value)
+	if err != nil {
+		return KeyValuePair{}, err
+	}
+	return KeyValuePair{Key: key, Value: value}, nil
+}
+func expandKvpsInPlace(kvps KeyValuePairs, expander SingleWordExpander) error {
+	for i, kvp := range kvps {
+		newKvp, err := expandKvp(kvp, expander)
+		if err != nil {
+			return err
+		}
+		kvps[i] = newKvp
+	}
+	return nil
+}
+
+func expandSliceInPlace(values []string, expander SingleWordExpander) error {
+	for i, v := range values {
+		newValue, err := expander(v)
+		if err != nil {
+			return err
+		}
+		values[i] = newValue
+	}
+	return nil
+}
+
+type EnvCommand struct {
+	BaseCommand
+	Env KeyValuePairs // kvp slice instead of map to preserve ordering
+}
+
+func (c *EnvCommand) Expand(expander SingleWordExpander) error {
+	return expandKvpsInPlace(c.Env, expander)
+}
+
+type MaintainerCommand struct {
+	BaseCommand
+	Maintainer string
+}
+
+type LabelCommand struct {
+	BaseCommand
+	Labels KeyValuePairs // kvp slice instead of map to preserve ordering
+}
+
+func (c *LabelCommand) Expand(expander SingleWordExpander) error {
+	return expandKvpsInPlace(c.Labels, expander)
+}
+
+type AddCommand struct {
+	BaseCommand
+	Srcs []string
+	Dest string
+}
+
+func (c *AddCommand) Expand(expander SingleWordExpander) error {
+	dst, err := expander(c.Dest)
+	if err != nil {
+		return err
+	}
+	c.Dest = dst
+	return expandSliceInPlace(c.Srcs, expander)
+}
+
+type CopyCommand struct {
+	BaseCommand
+	Srcs []string
+	Dest string
+	From string
+}
+
+func (c *CopyCommand) Expand(expander SingleWordExpander) error {
+	dst, err := expander(c.Dest)
+	if err != nil {
+		return err
+	}
+	c.Dest = dst
+	from, err := expander(c.From)
```

Comment:
I don't think this is supported atm. We have to be careful about this as it may prevent us from determining dependencies between stages before the image config has been pulled. We may want to do this in the context of global args like the strings in `FROM` command but that's a discussion for later.